METRON-830 Adding StringFunctions to Stellar - chop, prependifmissing, appendifmissing and countmatches

anandsubbu opened the pull request:
## Contributor Comments
Added stellar functions and unit tests for the following:
chop -> Remove the last character from a String.
prependifmissing -> Prepends the prefix to the start of the string if the string does not already start with any of the prefixes.
appendifmissing -> Appends the suffix to the end of the string if the string does not already end with any of the suffixes.
countmatches -> Counts how many times the substring appears in the larger string.


## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron (Incubating).  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  


In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?


### For code changes:
- [ ] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [ ] Have you included steps or a guide to how the change may be verified and tested manually?
- [ ] Have you ensured that the full suite of tests and checks have been executed in the root incubating-metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [NA] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  bin/generate-md.sh
  mvn site:site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommened that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

ottobackwards commented on metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/StringFunctions.java:
```diff
@@ -343,4 +343,89 @@ public Object apply(List<Object> args) {
+
+    @Override
+    public Object apply(List<Object> strings) {
+
```
Invalid Cast exception?

mattf-horton commented on metron-platform/metron-common/src/test/java/org/apache/metron/common/dsl/functions/StringFunctionsTest.java:
```diff
@@ -207,4 +207,78 @@ public void testFormatWithNoArguments() throws Exception {
+  public void testChop() throws Exception {
+    Assert.assertEquals("ab",   run("CHOP('abc')", new HashedMap()));
+    Assert.assertEquals(null,   run("CHOP(null)", new HashedMap()));
+    Assert.assertEquals("abc",  run("CHOP(msg)", ImmutableMap.of("msg", "abc\r\n")));
```
Why does this work, when the string "abc\r\n" is never processed by any of the shell or java string input processors that might convert it?

cestella replied:
Because it's coming in through a variable `msg`.  The idea being that it could be existent in a message flowing through the topologies.  Stellar can't represent `\n` and `\r`, but they can get in there honestly.

ottobackwards commented:
Thanks for the contribution!

Quick review, first thing please follow the style guide for naming classes.  You can see from the other stellar function classes.

It would be good to have some tests that aren't with string or null, to verify the behavior.

What happens if I pass in a number literal?

mattf-horton commented on metron-platform/metron-common/src/test/java/org/apache/metron/common/dsl/functions/StringFunctionsTest.java:
```diff
@@ -207,4 +207,78 @@ public void testFormatWithNoArguments() throws Exception {
+  @Test
+  public void testPrependIfMissing() throws Exception {
+    Assert.assertEquals("xyzabc",     run("PREPENDIFMISSING('abc', 'xyz')", new HashedMap()));
+    Assert.assertEquals("xyzXYZabc",  run("PREPENDIFMISSING('XYZabc', 'xyz', 'mno')", new HashedMap()));
```
Please add a case that exercises non-null alt prefix, eg:
```
Assert.assertEquals("mnoXYZabc",  run("PREPENDIFMISSING('mnoXYZabc', 'xyz', 'mno')", new HashedMap()));
```

anandsubbu replied:
Done, added.

mattf-horton commented on metron-platform/metron-common/src/test/java/org/apache/metron/common/dsl/functions/StringFunctionsTest.java:
```diff
@@ -207,4 +207,78 @@ public void testFormatWithNoArguments() throws Exception {
+  }
+
+  @Test(expected = ParseException.class)
+  public void testPrependIfMissingWithIncorrectArgs() throws Exception {
```
I may be wrong, but I think JUnit expected errors do not auto-continue inside an errored test case.  So, you're only testing the first sub-case.  

To fix, you either need 3 test cases, or you need to do try/catch/finally contexts around each `run()`, where the catch eats the exception but sets a marker flag, and the finally asserts the marker flag.

anandsubbu replied:
Yup, you're right. Added try/catch blocks now.

mattf-horton commented on metron-platform/metron-common/src/test/java/org/apache/metron/common/dsl/functions/StringFunctionsTest.java:
```diff
@@ -207,4 +207,78 @@ public void testFormatWithNoArguments() throws Exception {
+  }
+
+  /**
+   * APPENDIFMISSING StringFunction
```
Same issues here of course, and in COUNTMATCHES.

anandsubbu commented:
Hi @mattf-horton  and @ottobackwards , thank you for the review. Please see review comments addressed as a part of my latest commit, and also responses inline.

anandsubbu commented on metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/StringFunctions.java:
```diff
@@ -343,4 +343,89 @@ public Object apply(List<Object> args) {
+  @Stellar( name = "COUNTMATCHES"
+          , description = "Counts how many times the substring appears in the larger string"
+          , params = { "str - the CharSequence to check, may be null", "sub - the substring to count, may be null"}
+          , returns = "the number of occurrences, 0 if either CharSequence is null"
```
Corrected.

anandsubbu commented on metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/StringFunctions.java:
```diff
@@ -343,4 +343,89 @@ public Object apply(List<Object> args) {
+          break;
+        case 3: prefixed = org.apache.commons.lang3.StringUtils.prependIfMissing((String) strings.get(0), (String) strings.get(1), (String) strings.get(2));
+          break;
+        default: throw new IllegalArgumentException("[PREPENDIFMISSING] incorrect arguments. Usage: PREPENDIFMISSING <String> <prefix> [<prefix>...]");
```
Now included the arguments provided by user as well, apart from Usage info.

anandsubbu commented on metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/StringFunctions.java:
```diff
@@ -343,4 +343,89 @@ public Object apply(List<Object> args) {
+
+      String prefixed;
+      switch (strings.size()) {
+        case 2: prefixed = org.apache.commons.lang3.StringUtils.prependIfMissing((String) strings.get(0), (String) strings.get(1));
```
Changed the import statement to use lang3.

anandsubbu commented on metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/StringFunctions.java:
```diff
@@ -343,4 +343,89 @@ public Object apply(List<Object> args) {
+
+  @Stellar( name = "PREPENDIFMISSING"
+          , description = "Prepends the prefix to the start of the string if the string does not already start with any of the prefixes"
+          , params = { "str - The string.", "prefix - The prefix to prepend to the start of the string", "prefixes - Additional prefixes that are valid" }
```
Sure, done.

anandsubbu commented on metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/StringFunctions.java:
```diff
@@ -343,4 +343,89 @@ public Object apply(List<Object> args) {
+    }
+  }
+
+  @Stellar( name = "PREPENDIFMISSING"
```
Sure, changed.

anandsubbu commented on metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/StringFunctions.java:
```diff
@@ -343,4 +343,89 @@ public Object apply(List<Object> args) {
     }
   }
+
+  @Stellar( name="CHOP"
```
IMHO the `chop` function would be helpful when fine-tuning string extraction operations through stellar (E.g. when extracting specific variations of hostnames or URLs). 

Aside of that, I went ahead and added the `chomp` string function as well since I agree that it would be of use.

ottobackwards replied:
If we are going to do URL/URI work, we should create functions off of the URI / URL classes and use their specific parsing for uri named component parts

ottobackwards replied:
That is not to say the CHOP cannot be useful however

anandsubbu replied:
Okay, thanks. I stand corrected about the URI/URL.

cestella commented:
This looks good to me, anand.  Thanks so much for the contribution.  I'll spin it up and try the functions out in a REPL.  @mattf-horton @ottobackwards anything other outstanding concerns?

ottobackwards commented:
+1, thanks for the contribution

mattf-horton commented on metron-platform/metron-common/README.md:
```diff
@@ -167,6 +172,14 @@ The `!=` operator is the negation of the above.
+  * Input:
+    * string - The string to be appended.
+    * suffix - The string suffix to append to the end of the string.
+    * suffixes - Optional - Additional string suffixes that are valid terminators.
```
Please state the format of "suffixes".  Is it a string, a regex, an array of strings, or what?  If it is a single string, how does it represent plural suffixes?

anandsubbu replied:
Hi @mattf-horton , I was trying to following the same wording as the [apache commons documentation](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html) for PrependIfMissing, which uses the plural form. 

Let me know if it would be better if I changed the text as follows:

`* suffix - The string suffix to append to the end of the string.`
`* additionalsuffix - Optional - An additional string suffix that is a valid terminator.`

mattf-horton replied:
Ah, got it.  The documentation you point at for apache commons PrependIfMission method actually uses signature:
```
CharSequence... suffixes
```
so it is a varargs usage.  You can do the same with Stellar, or we have found it better to allow a list of strings instead (which you would explode into the calling varargs in the java code).  Using a list instead of varargs makes it much easier to extend the function later if we choose.

It's up to you whether to do this now.  We are trying to close a RC, so there is some time pressure.  Maybe you should just open a jira to enhance this later with a list arg.  

The current implementation of your call to PrependIfMissing is such that the "additionalsuffix" description you suggest would be correct.

anandsubbu replied:
Okay, I will retain additionalsuffix for now. Created METRON-847 to track the enhancement.

mattf-horton commented on metron-platform/metron-common/README.md:
```diff
@@ -434,6 +466,14 @@ The `!=` operator is the negation of the above.
+  * Input:
+    * string - The string to be prepended.
+    * prefix - The string prefix to prepend to the start of the string.
+    * prefixes - Optional - Additional string prefixes that are valid.
```
ditto

mattf-horton commented on metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/StringFunctions.java:
```diff
@@ -343,4 +343,89 @@ public Object apply(List<Object> args) {
+    public Object apply(List<Object> strings) {
+
+      if(strings.size() == 0) {
+        throw new IllegalArgumentException("[CHOP] missing argument: string to be chopped");
```
Any thoughts about why treat nulls and empty strings differently?

anandsubbu replied:
Do you think adding a check for null as follows, would be the right way?

```
 if(strings == null || strings.size() == 0) {
    throw new IllegalArgumentException("[CHOP] missing argument: string to be chopped");
}
```

mattf-horton replied:
I think you need 
```
if (strings == null || strings.size() == 0 || strings.get(0) == null || strings.get(0).length == 0) { ...
```

anandsubbu replied:
Thanks, made this change.

anandsubbu replied:
Hi @mattf-horton , with the above change where we check for `strings.get(0) == null || strings.get(0).toString().length() `, the following test results in a failure:

`Assert.assertEquals("abc",  run("CHOP(msg)", ImmutableMap.of("msg", "abc\r\n")));`

Pardon my limited understanding, but this is what I inferred:
- The `if` condition for null and 0 length check occurs twice.
- In the case of the `msg` test case where we check for `\n\r`, the initial contents of the `strings` is null and the variable substitution for `msg` occurs subsequently. So, this would end up failing for the 1st check. 

Due to this, I will modify the checks to only verify for `strings == null` and `string.size() == 0`. Please let me know if there is a better way this can be done.

mattf-horton commented on metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/StringFunctions.java:
```diff
@@ -343,4 +343,120 @@ public Object apply(List<Object> args) {
+          , params = {
+          "str - The string."
+          , "suffix - The string suffix to append to the end of the string"
+          , "suffixes - Optional - Additional string suffixes that are valid terminators"
```
Same doc concern as in the .md file: Is "suffixes" a string, a regex, an array of strings, or what? If it is a single string, how does it represent plural suffixes?

mattf-horton commented:
+1, thanks Anand!

anandsubbu commented:
Thanks a lot @cestella for your inputs and help. The latest commit addresses feedback from @mattf-horton and @cestella.

anandsubbu commented:
@mattf-horton - if you feel the latest commit addresses your feedback, can you please help with merging this pull request? I already have your +1 and @ottobackwards 's as well. Thank you!
